### ISSUE TITLE
Option to specify the 'temp/partial files' directory for a defined Folder

### ISSUE THREAD

**Issue by fragtion** [syncthing/syncthing#2208]
Hi. I've been experiencing good reliability syncing Maildirs with Syncthing for a few weeks already - however, from time to time an occasional blank mail may now show up in my IMAP folders/e-mail client. I suspect that this is due to the partial transfers and/or temp files being read by dovecot (perhaps locking them somehow could help with this?). I was able to ignore dovecot's index files in Syncthing using .stignore, but I can't seem to find any way to ignore Syncthing's temp files in Dovecot's configuration. Now I'm sure there could be many other situations where the temp files appearing in 'sensitive' sync folders (folders being continuously watched by other software) could cause similar problems too. If Syncthing had its own configuration option to configure a custom directory for storing temp files & partial transfers on a Folder-by-Folder basis (even if this is only configurable via the XML config) it might be incredibly useful in these cases. As far as I can tell it is not yet possible to do this? Perhaps someone knows how I can configure dovecot to play a little better with Syncthing? Or a more elegant solution? Designate maildir syncing tools like imapsync, offlineimap, isync, and dsync have not worked well for me - at least not as well as Syncthing... so it would seemingly be perfect, were it not for this side effect, which I imagine should not be too troublesome to work around? Thanks

**Comment by calmh** [comment]
Perhaps we could stick temp files into `.syncthing/` at the folder root or so; we need to have some space within the actual folder that we're allowed to write to, and that seems to be a place as good as any.

**Comment by fragtion** [comment]
Yes, at least for my case that would be perfect! It would be bonus if Syncthing could 'support' it if the user were to symlink that .syncthing folder to point to another folder elsewhere such as on another device, and this would also be at their own risk due to moving that temp folder not being officially supported

**Comment by calmh** [comment]
> to point to another folder elsewhere such as on another device

That's exactly the reason I don't want to make it too configurable. :) Doing that would break completely, as we can no longer rename the temp file to the actual target.

**Comment by uok** [comment]
@fragtion what exactly are you using Syncthing for?

I like that the tmp files are in the same directores as the finished files because it's a simple indicator (without opening Syncthing GUI) that a file is still in progress or if it stays there more than 1 day something went wrong, etc.

**Comment by fragtion** [comment]
@uok: Well I'm using Syncthing for a variety of different tasks actually :) But the issue I'm describing here pertains to using Syncthing for syncing Maildirs (as I have described in my first post here)

I agree that there is an element of convenience for the human to 'see' transferring files taking place in the actual target subfolder - but what I'm saying is that this does have a tendency to cause unwanted behavior in some scenarios where partial transfers/ temp files could be seen as alien to the other application which is actually using those synced folders, which is why I am requesting the option to allow for a more "non-destructive" approach, even if it is not the default behaviour

It could also be argued that it is more practical & easier to "clean up" one ".syncthing" folder when you decide to un-share it, than it is to search recursively for any potential remnants/traces from past failures. Plus, there's already a web GUI for seeing active transfers, right?...

Ideally it should be configurable, then we could choose either mode of operation depending what type of folder is being shared

**Comment by uok** [comment]
I read the first post, I just wanted to know because Dovecot runs on many servers worldwide and it also supports replication. Because of its large userbase there are tools for maildir sync (e.g. http://isync.sourceforge.net/) and two-way sync (e.g. http://wiki2.dovecot.org/Tools/Doveadm/Sync) that are designed specifically for your task.

**Comment by fragtion** [comment]
Thanks for your suggestions @uok 

As mentioned, I am aware of those tools (such as isync) and have already tried them without much satisfaction. I have also tried Dovecot's own built-in replication, and unfortunately experienced major issues with this. I was hoping to avoid going into the specifics as that is beyond the point of this request and is taking this off topic somewhat -- but since you have asked I will be happy to address them now. 

Baiscally we are hosting mail for many clients (domains) on our server and not all of them should be replicated. From my understanding, dovecot's replication configuration is an "all-or-nothing" approach, meaning that we can't specify individual replication servers per domain, or exclude certain domains or users from the built-in replication as the replication configuration seems to lack this flexibility

While dovecot does allow you to run the doveadm/dsync utilities independantly (via crontab for example), there is a bug with the more efficient "stateful" synchronization when doing it in this manner, where the replicas stop syncing and a 'full sync' needs to be run frequently order to return consistency. Doing this frequently causes unwanted traffic overheads on our network

As mentioned, Syncthing has finally proven to be the most effective solution to our maildir syncing needs, apart from this one specific issue. I really believe that this feature could come in handy for other users as well who prefer that there are no temp working files in the target directory, and a lack of such a feature could really be be a dealbreaker in these cases

**Comment by lfam** [comment]
Being able to specify the location of temporary files seems fine to me, but using Syncthing to perform multi-way sync of Maildirs sounds like a bad idea to me.

Yes, Maildirs are just files in the filesystem, but you are supposed to interact with them over an API, just like with git repos: https://en.wikipedia.org/wiki/Maildir#Technical_operation

Letting Syncthing copy files between machines in a random order is going to break your Maildirs.

This same question was asked on the mutt-users mailing list (they asked about Dropbox, but the same issues will probably apply):
http://permalink.gmane.org/gmane.mail.mutt.user/41461

I could see the vector clocks helping to some degree.

Have you tried to use isync / mbsync? It is designed to do exactly this, safely.

**Comment by fragtion** [comment]
I was able to get dovecot replication to work on a user-by-user basis by setting the mail_replica manually option for each user in the userdb, rather than specifying it as a global plugin variable in the main conf. It's quite a complex, mostly undocumented (and unconventional) setup. But even this official solution only supports master-master mode.

So I am still using Syncthing's UI for my 3rd node (my personal PC). I enjoy seeing Syncthing's status (web ui) so I know that everything is in identical sync; Plus, it just works perfectly (apart from the tmp files causing empty mails, and assuming dovecot's indexes/uid-caches are ignored from the sync).

I've tried isync, mbsync, offlineimap, etc and they all shared similar failings, such as having to re-copy entire mails (over the network) when the user moves them to another folder or trash, unable to properly renamed/delete mail folders, or needing to work via IMAP protocol (so UID's and filenames are not consistent among peers, or needing independent cache databases which either go out of sync or cause problems when folders are deleted).

I think that once Syncthing allows all of its files (including temp files) to operate under one .syncthing/ dir, and allows nuking of ignored files (#1631), then it will work even more perfectly for this, because it won't leave any of its own traces in the folders, and will fix the folder deletion problem

To me it's more an issue of performance than it is risk (and I'm prepared to take the risk as my own) - yes dovecot needs to re-index the folder each time it spots a foreign file, causing a readdir() each time the sync causes a change to the filesystem, and yes mails _should_ be delivered through the local MTA, but in theory any file in that directory is read and added to the index anyway, so if I've have had great success with this form of mail synchronization then I see no reason to stop doing it. I'd need a very practical proof/explanation to show why maintaining syncthing sync of two previously rsync'ed maildir replicas each controlled by independant dovecot sessions, poses any (practically) serious risks :)

**Comment by calmh** [comment]
All the maildir stuff aside, this

> allowing this minor setting to be configured could prove extremely useful for other users too who prefer that there are no temp working files in the target directory

sounds like it holds water to me. So this is a legit feature.

**Comment by c-cube** [comment]
@lfam sorry to disturb, but it looks to me like syncthing is ok for replicating a maildir in a read-only way (the server would have it as a master directory), doesn't it? It's only if local changes are to be propagated back to the server that race conditions might occur. Correct me if I'm wrong (I'm looking for a simple solution to have my mails offline on my laptop, and since I already use syncthing...).

**Comment by lfam** [comment]
@c-cube I use the mbsync program, distributed as isync by many package managers, to sync mail over IMAP.

**Comment by fragtion** [comment]
What would need to be changed in order to allow a custom temp file directory as described? I still desire this functionality and may have to fork if this is considered pointless for the main branch.. How extensive would the changes be? Thank you

**Comment by calmh** [comment]
grep for `tempNamer` or `defTempNamer` and you'll find the places where the temporary file names are generated

**Comment by fragtion** [comment]
Perfect, going to try. Thank you

**Comment by generalmanager** [comment]
Couldn't you just ln -s the path you actually want to use for the temp directory?

Am 9. November 2016 10:33:04 MEZ, schrieb Dimitri Pappas <notifications@github.com>:

**Comment by calmh** [comment]
Nope. We create temp files in actual destination directory, which is the problem they're looking to solve here.

**Comment by fragtion** [comment]
> Couldn't you just ln -s the path you actually want to use for the temp directory?
At the moment the tmp file is created in the same folder as the destination file, so I can't see how a symlink would help. There needs to be a way (hidden advanced option, perhaps?) to save the tmp files in a custom directory, and only move them to the final directory once the file is ready - that's is being requested

**Comment by calmh** [comment]
> hidden advanced option, perhaps?

I think that's the way, to start with.

**Comment by MRACHINI** [comment]
been waiting for this feature for a while, any idea when it would be available in stable ?

**Comment by calmh** [comment]
As you can see from the milestone, this is not planned. Anyone is welcome to contribute the feature, though.

**Comment by fragtion** [comment]
Looks like @svatasoiu has already figured this out from a coding perspective, but the change was committed to their fork. What would be the easiest/standard way of merging those changes with this original master branch? Would someone have to submit a pull request directly to this project for it to be considered? Thx

**Comment by AudriusButkevicius** [comment]
It's not as trivial. That doesn't handle crossing filesystem boundaries.

**Comment by calmh** [comment]
It will have to be a requirement that the temp dir is on the same filesystem as the folder or everything will fail. I hadn't actually seen that code, and I didn't really look at it now. A PR is a requirement, and a good start, yes.

**Comment by HermanEldering** [comment]
Just a short message to add another use case for this feature. I've just started experimenting with btrfs and snapper to create snapshots of my synced folders. This creates a sort of backup without storing duplicate data. However, because the temporary files are inside the same folder those are included in the snapshots and are thus retained for the lifetime of the snapshots. Ideally, I would store all temporary files inside a single folder. That folder could then be a btrfs subvolume, so that it would not be included in the snapshots. The move operation for completed files will still work in that case I think.

**Comment by fragtion** [comment]
Hi All

I see Jakob removed the "Unplanned" tag from this issue. Is this because it is not considered important enough to fall under that category anymore, or could it be because this is already being worked on for a future release? Of course I am hopeful that it is the second option, but if not, then I would like to fork the project in order to start working on a prototype of this desired functionality, as I still need it for my usage case.

May I please just request a pointer(s) of where to focus my efforts with this, as the project has a relatively large codebase compared to projects I normally work on. I'm guessing that "lib/model" should be left alone, and that focus should be in "/lib/fs/", especially "tempname.go" ? If someone who understands the project's operational framework structure could just guide me with some pointers to get started, I would greatly appreciate it. If I then fail to come up with an appropriate pull request, maybe someone else can still use the information as a head start. Thank you =)

**Comment by AudriusButkevicius** [comment]
No, this is not something that is being worked on. You'll have to tap into lib/model, as that is where the temporary file creation happens (somewhere in sharedPullerState), yet I suspect you'll have to touch other parts of the code base too.

**Comment by imsodin** [comment]
@fragtion Something else to consider: ".stfolder" is a directory now, so something like ".stfolder/temp-files" seems like a natural alternative location for temp files.

**Comment by fragtion** [comment]
Thanks for clarification on that, @AudriusButkevicius. I did notice that the temp files are basically identified by the 'TempPrefix' variable and that this is hard-coded to operate within the same sync folder, so the workaround would be adding a few lines of code everywhere to enforce the need for an additional folder

I think the best approach is to see if a custom temp path is defined in the advanced folder options, and if this field is not empty and the defined path is R+W accessible, then it writes the temp files there. If that field does not exist, is inaccessible, or is null/empty, then it reverts to the default operation.

Knowing that nobody else plans to implement this with any level of urgency, I will attempt it when I get a chance and advise of any progress on this thread. Any other contributors are also welcome to assist in the meanwhile

@imsodin That's an excellent point. Wouldn't it make the most sense to use .stfolder/temp-files as the default mode of operation then? Since it is a sub-folder of the sync folder anyway, it should in theory not cause any problems to do this, as we can assume that this folder will share the same space constraints as the sync folder itself. It seems to me that if anything, it should be alleviating potential issues further, and that the implementation of this feature is becoming more of a no-brainer as time moves on (note that we are now 2 years on since this issue was first raised). So, I think it would be easier for now to just change the hardcoding so that it uses this approach, instead of defining a custom folder for the temp files. The only caveat I can see, lies with long filename paths. I guess time will tell

**Comment by AudriusButkevicius** [comment]
There is more than just that, we now have this concept of filesystem abstraction which prevents doing any file operations outside of the folder, so if you wish this path to be adjustable outside of the folder root, then the change is a bit more involved.

**Comment by Ferroin** [comment]
Based on the first few comments, I think just having an option to drop the temporary files in the `.stfolder` marker folder instead of having them in-line alongside the regular files would probably cover most of the cases where this actually matters beyond simple user preference (that is, it would be enough to deal with software that doesn't have a way to ignore them).

**Comment by fragtion** [comment]
Just want to follow up on this issue. Does syncthing still create the tmp files in the target save directory? Realistically, would changing the temp files folder to .stfolder require a lot of changes to the codebase? Thank you

**Comment by calmh** [comment]
It still does, as the issue is not closed. The presence of the "good first issue" tag indicates that we thought at some point that this should be a simple enough thing to change if someone sat down to do it.

**Comment by fragtion** [comment]
Great pointers, thanks @calmh.

So it turns out that a one-liner edit in [TempName](https://github.com/syncthing/syncthing/blob/add12b43aa0bdf5e67d8f421c57a5ecafb3d25fa/lib/fs/tempname.go) was enough to achieve a working proof-of-concept:
-
`tdir := filepath.Dir(name)`
+
`tdir := ".stfolder/"`

I don't know much golang yet, so not sure if it would automatically substitute the forward slash for non-unix systems (in my case I only need this for a unix-based purpose so that doesn't matter); Also whether just a single folder for all tmp files could now lead to other "conflicts" arising where two different Folders might actively be transferring different files sharing the same name. With that in mind, I did try appending "filepath.Dir(name)" to ".stfolder/", but it crashed because that folder obviously didn't exist. Maintaining a parallel folder structure like this makes little practical sense though I suppose, so a better approach for avoiding collisions might be to append a random string or timestamp to the TempPrefix (".syncthing."/"~syncthing~"). Should be a safe and effective way to deal with that concern, as far as I can tell

**Comment by imsodin** [comment]
The conflict problem could also be solved, by using the full relative part (place `/foo/bar/baz` in `.stfolder/tempfiles/foo/bar/baz`), at the cost of some overhead due to handling the directories.

This is meant as an optional feature, right? If yes, I'd be in favor of the current behaviour as default. In that case it should be fine to ignore the multiple filesystem issue.

**Comment by mfonsos** [comment]
Hello guys,

I am trying to follow your conversation...is it on config.xml file that I have to enter a hardcoded value or edit the source file and do it manually...I would appreciate if someone points me to the right direction...thank you!

**Comment by AudriusButkevicius** [comment]
You'd have to modify and build your own fork of syncthing. The change to the code can be seen in the commits above that reference this issue, and the build instructions are on the docs website.

**Comment by fergalish** [comment]
> Another issue some people will run into is where the folder is in fact spread over multiple filesystems and we can't rename from the .stfolder into the final resting place of the file. I bet that's fairly unusual though.

Hi @calmh, I have exactly this scenario.
I recompiled with @fragtion's commit above, and changed the code in lib/fs/basicfs.go as below.
Short version: use copy-and-delete instead of rename.

Waiting to see what the performance hit will be like.

```
func (f *BasicFilesystem) Rename(oldpath, newpath string) error {
        oldpath, err := f.rooted(oldpath)
        if err != nil {
                return err
        }
        newpath, err = f.rooted(newpath)
        if err != nil {
                return err
        }
//      return os.Rename(oldpath, newpath)

    inputFile, err := os.Open(oldpath)
    if err != nil {
        return fmt.Errorf("Couldn't open source file: %s", err)
    }
    outputFile, err := os.Create(newpath)
    if err != nil {
        inputFile.Close()
        return fmt.Errorf("Couldn't open dest file: %s", err)
    }
    defer outputFile.Close()
    _, err = io.Copy(outputFile, inputFile)
    inputFile.Close()
    if err != nil {
        return fmt.Errorf("Writing to output file failed: %s", err)
    }
    // The copy was successful, so now delete the original file
    err = os.Remove(oldpath)
    if err != nil {
        return fmt.Errorf("Failed removing original file: %s", err)
    }
    return nil
}
```

**Comment by AudriusButkevicius** [comment]
That's not atomic, so can lead to corruption.

**Comment by fergalish** [comment]
> That's not atomic, so can lead to corruption.

Good point.
My use case is to synchronize an on-the-fly encrypted mapping of a plaintext directory (encfs --reversewrite). But encfs doesn't allow arbitrary filenames to be written to the encrypted directory, it must match the block cipher and produce sensible plaintext filenames.
Therefore files like ".stfolder" ".stignore" and ".syncthing.filename.tmp" in the encrypted mapping give an i/o error from encfs, because they don't decrypt.

One option is to see if encfs can ignore the invalid filenames, but I think that'll not be possible - the encrypted directory is only a mapping and doesn't have underlying physical storage. Only the plaintext has a physical storage, and if encfs can't decrypt the filename or its contents how can it know what to store?

Alternative solution would be to not use temporary filenames while syncing (i.e. use the actual "filename" instead of ".syncthing.filename.tmp"

**Comment by calmh** [comment]
I can say with confidence that neither of those options will be implemented "as is" (copying in from the temp to the final file, and not having a temp file to begin with). There is some though around another way to do in place modifications, but it's complex to do correctly.

If this is implemented, it will be a sync failure to not be able to rename the temp file to the final file, such as if they are on different filesystems.

**Comment by TCB13** [comment]
> Yes, Maildirs are just files in the filesystem, but you are supposed to interact with them over an API, just like with git repos: https://en.wikipedia.org/wiki/Maildir#Technical_operation
> 
> Letting Syncthing copy files between machines in a random order is going to break your Maildirs.

In my experience as long as you don't sync `dovecot.*` files such as indexes, logs, cache etc the files that store the actual emails will eventually sync / other instances of dovecot will eventually pick them.

**Comment by fragtion** [comment]
Yup. But that is not enough unfortunately, as the other thing you want to look out for is the 'new/' folder. If you have two or more IMAP clients connecting to independent IMAP server nodes, the any new mails will be moved to the 'cur/' folder with different names on each server, leading to duplicate formation in 'cur' folder. I managed to effectively work around that problem by removing 'new/' functionality from dovecot's Maildir library, and then in order to keep other MTA deliveries working, symlink 'new/' to 'cur/' for each folder. You're welcome to see my forks https://github.com/fragtion/syncthing and https://github.com/fragtion/dovecot-core for more details. It's been working great for me for some years now already.

**Comment by AudriusButkevicius** [comment]
Please, keep this on subject and talk dovecote specifics via some other medium.

**Comment by fragtion** [comment]
Hey all

Sorry for bumping an old thread.  I'm actually just facing a similar scenario to my opening issue here, but with a slight twist. Also can't seem to find much by way of related issues, so I thought I'd just expand this issue with a possible similar idea/suggestion

Ok so this time I'm trying to do a more time-sensitive operation (an after effects 'multi-machine' render accross the network). Here it would be extremely helpful if either: 
1. there is an option to skip creation of the tmp file altogether, so that the file being written to during sync/transfer, IS the final output destination filename in the folder
or
2. we can enable an option whereby syncthing creates a temporary placeholder/"lock" file in the final destination folder, which has the SAME NAME, until being replaced with the tmp file once the transfer is complete

The reasons for this are twofold:
1. even with inotify/filesystem watchers, there always seems to be some delay before the local device becomes aware of a filesystem change and notifies other devices. I can think of many reasons why it would be advantageous for the remote nodes to establish presence of any new files on remote nodes where t = after the file was created on the first node, and t = before the transfer of the actual file content completes. it may not be ideal from an IO standpoint, but can help mitigate against conflicts
2. apart from just mitigating against conflicts, other software on remote devices (the aerender process, in my case here) would significantly benefit from a more real-time state of which files *should* be there, even if the state is *still* not technically realtime/synchronous.

Again, I don't think this should necessarily be default behavior.  But it would certainly be a useful feature to have optional support for at this stage perhaps?

Anyway thanks for your time in reading/considering this :)

**Comment by tysoncung** [comment]
I'm interested in working on this issue. Could you provide more context about what you're looking for? Any additional details about requirements or constraints would be helpful.
